src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -41,54 +41,92 @@ module.exports = function<T, P, I, TI, C, CX>(
     contextDepth: number
   };
 
-  function createContainerState(rootInstance) {
-    return {
-      rootInstance,
-      contextFibers: null,
-      contextValues: null,
-      contextDepth: -1,
-    };
-  }
+  // Current container context on the stack.
+  let rootInstance : C | null = null;
+  let contextFibers : Array<Fiber | null> | null = null;
+  let contextValues : Array<CX | null> | null = null;
+  let contextDepth : number = -1;
 
-  // If we meet any portals, we'll go deeper.
-  let containerStack : Array<ContainerState> = [];
-  let containerDepth : number = -1;
+  // If we meet any portals, we'll pack outer context in this array.
+  let portalStack : Array<ContainerState> = [];
+  let portalDepth : number = -1;
 
   function getRootHostContainer() : C {
-    if (containerDepth === -1) {
-      throw new Error('Expected to find a root container.');
+    if (rootInstance == null) {
```

Comment:
For the new Context API I was thinking that we should require a default value so that the type is always guaranteed to be satisfied and the user doesn't have to add extra null checks. Maybe we should do the same for host context?